### ISSUE TITLE
Auto-profile launcher won't switch profiles if controller is turned on in-game

### ISSUE THREAD

**Issue by CorpusBava** [Ryochan7/DS4Windows#578]
In latest build 1.6.10 (haven't tested earlier builds), auto-switching profile function will not work if I'm already in game and turn my DS4 controller on.  Most of the time I forget to turn my controller on before I've launched a game or program that I need to use my controller with.  In the original builds of DS4Windows auto-switching of profiles always worked even if I turned my controller on while in game.  Now it just stays on whatever profile I last activated in DS4Windows.  Is there a fix for this?

**Comment by Ryochan7** [comment]
The status of the problem has not changed since at least issue #480 back in November 2018. The fix is a simple one to implement but any attempted fix has introduced application slowdown. Because of that, the issue has remained unfixed.

**Comment by Yohoki** [comment]
The easiest solution for this is to Alt+tab to another screen, wait a second and then go back to your game. It should detect that you've switched active windows and perform an auto-profile switch. For some reason it doesn't do this active window check on startup of the ds4windows application or when the controller is first plugged in. This becomes especially annoying if your controller becomes unplugged during gameplay, as you must repeat this process after plugging it back in. It's a major issue for those of us using USB mode.

Hopefully, Ryochan is able to find a fix that doesn't impact performance.

**Comment by CorpusBava** [comment]
> The easiest solution for this is to Alt+tab to another screen, wait a second and then go back to your game. It should detect that you've switched active windows and perform an auto-profile switch. For some reason it doesn't do this active window check on startup of the ds4windows application or when the controller is first plugged in. This becomes especially annoying if your controller becomes unplugged during gameplay, as you must repeat this process after plugging it back in. It's a major issue for those of us using USB mode.
> 
> Hopefully, Ryochan is able to find a fix that doesn't impact performance.

Yes, that's what I've been doing.  Alt-tabbing to recaptures focus.  I was hoping to automate the process with an autohotkey script or batch file of some kind but didn't have much success.  The inability to auto-launch issue with all later builds of DS4Windows is certainly not as crucial as something like, say, the lack of a reliable Exclusive Mode, but it would just be nice to have this feature back.  I'm fortunate in that my rig is fairly powerful so I don't really notice the slow-down issue but I understand others may.

**Comment by Ryochan7** [comment]
I thought I almost had something last night but it didn't work out when implementing the same checks in the ControlService Start routine. This whole issue started due to a design change made that delayed loading profiles until devices are detected.

With the Jays2Kings version, assigned profiles were loaded upon program startup before devices were checked. The ControlService Start and Hotplug routines were not responsible for loading profiles but just doing a simple check if an XInput controller needed to be registered. As it is now, the auto profile is actually loaded for controllers when an assigned application is detected but the Hotplug routine ends up loading the old selected profile anyway. Adding additional checks to skip profile loading and reset some dependent properties cause unwanted slowdown.

I will give this another go later. The auto profile routine will probably have to be changed to hopefully make external profile checks easier to handle. Editing DS4Form has always been trouble regarding changes introducing slowdown.

**Comment by Ryochan7** [comment]
Ended up changing how temp profiles are checked outside of DS4Form and that seemed to work out. Auto profiles will be loaded for recently connected controllers.

**Comment by CorpusBava** [comment]
Hey, great news!  I'll check it out when it becomes available.  Right now I'm using InputMapper with HIDGuardian despite the fact that I dislike so many things about it but it does do one thing exactly right that DS4Windows doesn't do - a fully working exclusive mode 100% of the time.  I had too many games/programs complaining about exclusive mode so I finally just had to chuck DS4Windows for the time being.  Retroarch for example would get completely confused detecting both 360 & DS4 controllers sometimes that the controller just refused to work under DS4Windows... most of the time.  That was finally what made it necessary to switch.   But it's great that you got the auto-profiles thing sorted.  That's definitely something InputMapper v1.6 can't do.  In fact, its auto-profile or even just controller recognition feature is non-functioning during controller hot swaps.

**Comment by Ryochan7** [comment]
InputMapper is still in sort of a rough spot. 1.6 has not been maintained in a long time but 1.7 still has not had an official stable release. Luckily 1.7 seems to be fairly close to a true public release and I think it is now supposedly feature complete according to what has been said in the YouTube videos. It is best to judge InputMapper by what is possible in the 1.7 betas.

With the IM installer, getting HidGuardian working with InputMapper would be much easier. DS4Windows would be able to use the same setup once it is installed though. I don't use HidGuardian regularly so I try to keep in mind what software is known to cause exclusive mode to fail. The old exclusive mode trick works for me most of the time but HidGuardian is a more reliable way of obtaining exclusive access to a device.

**Comment by CorpusBava** [comment]
Thanks for the info.  I tried 1.7 of IM but couldn't make heads nor tails of its plugin nature so maybe I'll just wait for the official release.  However, I always much preferred DS4Windows which has a very refined and more feature filled interface compared to IM, and I like that you didn't change anyone in that area over the years.  I'd drop IM in a minute and go back to DS4Windows if I could just get Exclusive Mode working 100% of the time.  Is there any way to get DS4Windows working with HIPGuardian so that it has that 100% success rate with Exclusive Mode like IM?  IM 1.6 is virtually useless for auto-switching of profiles if you have it enabled because it has a bug that causes your DS4 LED to burn at 100% full brightness until you click outside the IM window.  This burns through my battery within half a day.  So now I'm forced to use Xpadder in conjunction with IM to get autoprofles working right.  Would much prefer to have a single re-mapping/Xbox controller emulator program like DS4Windows

**Comment by Ryochan7** [comment]
There used to be an application for installing HidGuardian and getting the device list written to the registry. That application does not work anymore though. It relied on pulling data from my old web host that recently got shut down because I couldn't afford to keep it running. I will have to check if I still have the old source code around and maybe change it.

For now, there are some instructions for getting HidGuardian installed and configured. It is a bit of an undertaking.

https://github.com/Ryochan7/DS4Windows/wiki/DS4Windows-and-HIDGuardian---Install-and-Setup-Guide

**Comment by CorpusBava** [comment]
Thanks.  I was searching all over the web for instructions for getting HIDGuardian configured with DS4Windows and had no luck.  I really want to go back to DS4Windows because I really think it's the best Xbox controller emulator/re-mapper around.  Thanks for continuing to maintain it and for addressing any issues we may have with it.

**Comment by Ryochan7** [comment]
I changed the code for the old test installer and it does work with the latest changes. It links to the latest archive from vigem.org. The installer will install the HidGuardian driver and write an appropriate affected devices entry in the registry. It also works as an installer of ViGEmBus.

https://drive.google.com/file/d/1pHLtOeD_Qfo3nt9Pa3jYWo5cc5C3iuGk/view?usp=sharing

**Comment by CorpusBava** [comment]
Gresat!  So it's just a one step process now?  I was going to wait for the new DS4Windows build with the auto-profile fixes to configure HIDGiardian because it seemed so involved a process but maybe I'll give it a look now.  So I'm assuming I should probably uninstall my current version of HIDGuardian then use your new installer.  I wanted to mention something for the benefit of others who may be reading this topic and may wish to try this out.  I discovered that HIDGuardian can be a dangerous program if you use a third-party Windows uninstallation program like Total Uninstall Pro.  Using that program to uninstall HIDGuardian caused it to wipe some of my hardware keys in the Windows Registry which resulted in my system being unable to boot.  Just thought I'd warn others about this.

**Comment by Ryochan7** [comment]
You do have to be careful how you uninstall HidGuardian. The official instructions on the ViGEm site warn about potential problems. Never had any hardware keys get removed. While testing odd cases for the manual uninstall, I lost keyboard and mouse support entirely; I used a bad value for UpperFilters on purpose to see what would happen. I was able to restore the UpperFilters registry key via a recovery disk to fix that problem so I didn't have to resort to reinstalling Windows.

The program should be able to detect the currenly installed instance of HidGuardian and give an option to uninstall it. I did not test the ViGEm portion of the program so that part of the program should probably be ignored for now. This test driver installer program was originally made to experiment with WPF and also be used as a future replacement for the current Welcome dialog window used to install ScpVBus. Since I decided to not push ViGEmBus usage in the main build, this program has not been worked on much since early 2018.
